INTRODUCTION.md

```diff
@@ -388,6 +392,244 @@ may be produced out of order, since a sub-sequent message in a sub-sequent
+the PID's state has expired (due to topic retention, DeleteRercords,
+or compaction).
+
+The Java producer added quite a bit of error handling for this case,
+extending the ProduceRequest protocol to return the logStartOffset
+to give the producer a chance to differentiate between an actual
+UNKNOWN_PRODUCER_ID or topic retention having deleted the last
+message for this producer (effectively voiding the Producer ID cache).
+This workaround proved to be error prone (see explanation in KIP-360)
+when the partition leader changed.
+
+KIP-360 suggests removing this error checking in favour of failing fast,
+librdkafka follows suite.
+
+
+If the response is for the first ProduceRequest in-flight
+and there are no messages waiting to be retried nor any ProduceRequests
+unaccounted for, then the error is ignored and the epoch is incremented,
+this is likely to happen for an idle producer who's last written
+message has been deleted from the log, and thus its PID state.
+Otherwise the producer raises a fatal error
+(RD_KAFKA_RESP_ERR_UNKNOWN_PRODUCER_ID) since the delivery guarantees can't
+be satisfied.
+
+
+**Java Producer behaviour:**
+Retries the send in some cases (but KIP-360 will change this).
+Not a fatal error in any case.
+
+
+##### Standard errors
+
+All the standard Produce errors are handled in the usual way,
+permanent errors will fail the messages in the batch, while
+temporary errors will be retried (if retry count permits).
+
+If a permanent error is returned for a batch in a series of in-flight batches,
+the sub-sequent batches will fail with
+RD_KAFKA_RESP_ERR_OUT_OF_ORDER_SEQUENCE_NUMBER since the sequence number of the
+failed batched was never written to the topic log and next expected sequence
+thus not incremented on the broker.
+
+A fatal error (RD_KAFKA_RESP_ERR__GAPLESS) is raised to satisfy the gap-less
+guarantee (if `enable.gapless.guarantee` is set) by failing all queued messages.
+If there's community demand, a future configuration property will be added
+to change this behaviour to continue producing.
+
+
+##### Message persistance status
+
+To help the application decide what to do in these error cases, a new
+per-message API is introduced, `rd_kafka_message_persistance_status()`,
+which returns one of the following values:
+
+ * `RD_KAFKA_MSG_STATUS_NOT_PERSISTED` - the message has never
+   been transmitted to the broker, or failed with an error indicating
+   it was not written to the log.
+   Application retry will risk ordering, but not duplication.
+ * `RD_KAFKA_MSG_STATUS_POSSIBLY_PERSISTED` - the message was transmitted
+   to the broker, but no acknowledgement was received.
+   Application retry will risk ordering and duplication.
+ * `RD_KAFKA_MSG_STATUS_PERSISTED` - the message was written to the log by
+   the broker and fully acknowledged.
+   No reason for application to retry.
+
+This method should be called by the application on delivery report error.
```

Comment:
In the bindings I think this means we will now have a new error type `MessageError` which is a subclass of `Error` that adds an additional enum property `Status`.

Reply:
Maybe for .NET, which is the only client using different classes for succeeded or failed messages.

While the status will always be PERSISTED for successfully delivered messages, I don't think it hurts to provide a Status() method on the base producer message class so it is usable both for succeeded and failed messages. I could see a future enhancement which sets POSSIBLY_PERSISTED if producing with acks=1.

Reply:
yep, status properly belongs a level above error, you're right.